src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -259,58 +256,49 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
-            return [
+            [
                 [kernelConnection.id, JupyterNotebookView],
                 [`${kernelConnection.id} (Interactive)`, InteractiveWindowView]
-            ].map(([id, viewType]) => {
-                const controller = new VSCodeNotebookController(
-                    kernelConnection,
-                    id,
-                    viewType,
-                    label,
-                    this.notebook,
-                    this.commandManager,
-                    this.kernelProvider,
-                    this.preferredRemoteKernelIdProvider,
-                    this.context,
-                    this.pathUtils,
-                    this.disposables,
-                    this.languageService,
-                    this.workspace,
-                    this.isLocalLaunch ? 'local' : 'remote',
-                    this.interpreterPackages,
-                    this.configuration,
-                    this.widgetCoordinator
-                );
-                // Hook up to if this NotebookController is selected or de-selected
-                controller.onNotebookControllerSelected(
-                    this.handleOnNotebookControllerSelected,
-                    this,
-                    this.disposables
-                );
-
-                // We are disposing as documents are closed, but do this as well
-                this.disposables.push(controller);
-
-                return controller;
-            });
+            ]
+                .filter(([id]) => !this.registeredControllers.has(id))
+                .forEach(([id, viewType]) => {
+                    const controller = new VSCodeNotebookController(
+                        kernelConnection,
+                        id,
+                        viewType,
+                        label,
+                        this.notebook,
+                        this.commandManager,
+                        this.kernelProvider,
+                        this.preferredRemoteKernelIdProvider,
+                        this.context,
+                        this.pathUtils,
+                        this.disposables,
+                        this.languageService,
+                        this.workspace,
+                        this.isLocalLaunch ? 'local' : 'remote',
+                        this.interpreterPackages,
+                        this.configuration,
+                        this.widgetCoordinator
+                    );
+                    // Hook up to if this NotebookController is selected or de-selected
+                    controller.onNotebookControllerSelected(
+                        this.handleOnNotebookControllerSelected,
+                        this,
+                        this.disposables
+                    );
+
+                    // We are disposing as documents are closed, but do this as well
+                    this.disposables.push(controller);
+                    this.registeredControllers.set(controller.id, controller);
```

Comment:
once created we add it here, rather than returning it and trhying to add it elsewhere..